File: rust/tvm/src/ir/module.rs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

use std::collections::HashMap;
use std::iter::FromIterator;
use std::path::Path;

use thiserror::Error;
use tvm_macros::Object;

use crate::runtime::array::Array;
use crate::runtime::function::Result;
use crate::runtime::map::Map;
use crate::runtime::string::String as TVMString;
use crate::runtime::{external, IsObjectRef, Object, ObjectRef};

use super::expr::GlobalVar;
use super::function::BaseFunc;
use super::source_map::SourceMap;
use super::{relay, ty::GlobalTypeVar, ty::TypeData};

#[derive(Error, Debug)]
pub enum Error {
    #[error("{0}")]
    IO(#[from] std::io::Error),
    #[error("{0}")]
    TVM(#[from] crate::runtime::Error),
}

#[repr(C)]
#[derive(Object, Debug)]
#[ref_name = "IRModule"]
#[type_key = "IRModule"]
pub struct IRModuleNode {
    pub base: Object,
    pub functions: Map<GlobalVar, BaseFunc>,
    pub type_definitions: Map<GlobalTypeVar, TypeData>,
    pub source_map: SourceMap,
    // TODO(@jroesch): this is missing some fields
}

external! {
    // Parser functions
    #[name("relay.parser.ParseModule")]
    fn parse_module(file_name: TVMString, source: TVMString) -> IRModule;
    #[name("relay.parser.ParseExpr")]
    fn parse_expression(file_name: TVMString, source: TVMString) -> IRModule;
    #[name("ir.IRModule")]
    fn module_new(funcs: Map<GlobalVar, BaseFunc>, types: Map<GlobalTypeVar, TypeData>, attrs: Map<TVMString, ObjectRef>, global_infos: Map<TVMString, Array<ObjectRef>>) -> IRModule;
    // Module methods
    #[name("ir.Module_Add")]
    fn module_add(module: IRModule, type_name: GlobalVar, expr: BaseFunc, update: bool) -> IRModule;
    #[name("ir.Module_AddDef")]
    fn module_add_def(module: IRModule, type_name: GlobalTypeVar, type_data: TypeData, update: bool) -> ();
    #[name("ir.Module_GetGlobalVar")]
    fn module_get_global_var(module: IRModule, name: TVMString) -> GlobalVar;
    #[name("ir.Module_GetGlobalVars")]
    fn module_get_global_vars(module: IRModule) -> Array<GlobalVar>;
    #[name("ir.Module_Lookup")]
    fn module_lookup(module: IRModule, var: GlobalVar) -> BaseFunc;
    #[name("ir.Module_Lookup_str")]
    fn module_lookup_str(module: IRModule, name: TVMString) -> BaseFunc;
    #[name("ir.Module_GetGlobalTypeVars")]
    fn module_get_global_type_vars(module: IRModule) -> Array<GlobalTypeVar>;
    #[name("ir.Module_ContainGlobalVar")]
    fn module_contains_global_var(module: IRModule, name: TVMString) -> bool;
    #[name("ir.Module_ContainGlobalTypeVar")]
    fn module_contains_global_type_var(module: IRModule, name: TVMString) -> bool;
    #[name("ir.Module_LookupDef")]
    fn module_lookup_def(module: IRModule, global: GlobalTypeVar) -> TypeData;
    #[name("ir.Module_LookupDef_str")]
    fn module_lookup_def_str(module: IRModule, global: TVMString) -> TypeData;
    #[name("ir.Module_LookupTag")]
    fn module_lookup_tag(module: IRModule, tag: i32) -> relay::Constructor;
    #[name("ir.Module_FromExpr")]
    fn module_from_expr(expr: relay::Expr, funcs: Map<GlobalVar, BaseFunc>, types: Map<GlobalTypeVar, TypeData>) -> IRModule;
    #[name("ir.Module_Import")]
    fn module_import(module: IRModule, path: TVMString);
    #[name("ir.Module_ImportFromStd")]
    fn module_import_from_std(module: IRModule, path: TVMString);
}

// Note: we don't expose update here as update is going to be removed.

impl IRModule {
    pub fn new<'a, F, T, A, G>(funcs: F, types: T, attrs: A, global_infos: G) -> Result<IRModule>
    where
        F: IntoIterator<Item = (&'a GlobalVar, &'a BaseFunc)>,
        T: IntoIterator<Item = (&'a GlobalTypeVar, &'a TypeData)>,
        A: IntoIterator<Item = (&'a TVMString, &'a ObjectRef)>,
        G: IntoIterator<Item = (&'a TVMString, &'a Array<ObjectRef>)>,
    {
        module_new(
            Map::from_iter(funcs),
            Map::from_iter(types),
            Map::from_iter(attrs),
            Map::from_iter(global_infos),
        )
    }

    pub fn empty() -> Result<IRModule> {
        let funcs = HashMap::<GlobalVar, BaseFunc>::new();
        let types = HashMap::<GlobalTypeVar, TypeData>::new();
        let attrs = HashMap::<TVMString, ObjectRef>::new();
        let global_infos = HashMap::<TVMString, Array<ObjectRef>>::new();
        IRModule::new(
            funcs.iter(),
            types.iter(),
            attrs.iter(),
            global_infos.iter(),
        )
    }

    pub fn parse<N, S>(file_name: N, source: S) -> Result<IRModule>
    where
        N: Into<TVMString>,
        S: Into<TVMString>,
    {
        parse_module(file_name.into(), source.into())
    }

    pub fn parse_file<P: 'static + AsRef<Path>>(
        file_path: P,
    ) -> std::result::Result<IRModule, Error> {
        let file_path = file_path.as_ref();
        let file_path_as_str = file_path.to_str().unwrap().to_string();
        let source = std::fs::read_to_string(file_path)?;
        let module = IRModule::parse(file_path_as_str, source)?;
        Ok(module)
    }

    pub fn add<F>(&mut self, var: GlobalVar, func: F) -> Result<IRModule>
    // todo(@jroesch): can we do better here? why doesn't BaseFunc::Object work?
    where
        F: IsObjectRef,
        F::Object: AsRef<<BaseFunc as IsObjectRef>::Object>,
    {
        module_add(self.clone(), var, func.upcast(), true)
    }

    pub fn add_def(
        &mut self,
        type_name: GlobalTypeVar,
        type_data: TypeData,
        update: bool,
    ) -> Result<()> {
        module_add_def(self.clone(), type_name, type_data, update)
    }

    pub fn get_global_var<S>(&self, name: S) -> Result<GlobalVar>
    where
        S: Into<TVMString>,
    {
        module_get_global_var(self.clone(), name.into())
    }

    pub fn get_global_vars(&self) -> Result<Array<GlobalVar>> {
        module_get_global_vars(self.clone())
    }

    pub fn lookup(&self, var: GlobalVar) -> Result<BaseFunc> {
        module_lookup(self.clone(), var)
    }

    pub fn lookup_str<S>(&self, name: S) -> Result<BaseFunc>
    where
        S: Into<TVMString>,
    {
        module_lookup_str(self.clone(), name.into())
    }

    pub fn get_global_type_vars(&self) -> Result<Array<GlobalTypeVar>> {
        module_get_global_type_vars(self.clone())
    }

    pub fn contains_global_var<S: Into<TVMString>>(&self, name: S) -> Result<bool> {
        module_contains_global_var(self.clone(), name.into())
    }

    pub fn contains_global_type_var<S: Into<TVMString>>(&self, name: S) -> Result<bool> {
        module_contains_global_type_var(self.clone(), name.into())
    }

    pub fn lookup_def(&self, global: GlobalTypeVar) -> Result<TypeData> {
        module_lookup_def(self.clone(), global)
    }

    pub fn lookup_def_str<S>(&self, global: S) -> Result<TypeData>
    where
        S: Into<TVMString>,
    {
        module_lookup_def_str(self.clone(), global.into())
    }

    pub fn lookup_tag(&self, tag: i32) -> Result<relay::Constructor> {
        module_lookup_tag(self.clone(), tag)
    }

    pub fn from_expr<E>(expr: E) -> Result<IRModule>
    where
        E: IsObjectRef,
        E::Object: AsRef<<relay::Expr as IsObjectRef>::Object>,
    {
        Self::from_expr_with_items(expr, HashMap::new(), HashMap::new())
    }

    pub fn from_expr_with_items<'a, E, F, T>(expr: E, funcs: F, types: T) -> Result<IRModule>
    where
        F: IntoIterator<Item = (&'a GlobalVar, &'a BaseFunc)>,
        T: IntoIterator<Item = (&'a GlobalTypeVar, &'a TypeData)>,
        E: IsObjectRef,
        E::Object: AsRef<<relay::Expr as IsObjectRef>::Object>,
    {
        module_from_expr(expr.upcast(), Map::from_iter(funcs), Map::from_iter(types))
    }

    pub fn import<S: Into<TVMString>>(&mut self, path: S) -> Result<()> {
        module_import(self.clone(), path.into())
    }

    pub fn import_from_std<S: Into<TVMString>>(&mut self, path: S) -> Result<()> {
        module_import_from_std(self.clone(), path.into())
    }
}

#[cfg(test)]
mod tests {
    use super::relay::*;
    use super::*;
    use crate::ir::span::Span;
    use crate::ir::ty::{GlobalTypeVar, TypeData, TypeKind};
    use tvm_rt::IsObjectRef;

    fn add_dummy_functions(names: Vec<&str>) -> Result<IRModule> {
        let mut module = IRModule::empty()?;
        let x = Var::static_tensor("x".into(), vec![1, 1], DataType::float32());
        let params = vec![x.clone()];
        let func = relay::Function::simple(params, x);

        for name in names {
            let gv = GlobalVar::new(name.into(), Span::null());
            module = module.add(gv, func.clone())?;
        }

        Ok(module)
    }

    fn add_dummy_types(names: Vec<&str>) -> Result<IRModule> {
        let mut module = IRModule::empty()?;

        for name in names {
            let name: String = name.into();
            let name = GlobalTypeVar::new(name, TypeKind::Type, Span::null());
            let type_data = TypeData::new(name.clone(), vec![], vec![], Span::null());
            module.add_def(name, type_data, true)?;
        }

        Ok(module)
    }

    #[test]
    fn test_module_add() -> anyhow::Result<()> {
        let mut module = IRModule::empty()?;
        let x = Var::static_tensor("x".into(), vec![1, 1], DataType::float32());
        let params = vec![x.clone()];
        let func = relay::Function::simple(params, x);
        let module = module.add(GlobalVar::new("foo".into(), Span::null()), func)?;
        let lfunc = module.lookup_str("foo")?;
        let lfunc = lfunc.downcast::<relay::Function>()?;
        assert_eq!(lfunc.params.len(), 1);
        Ok(())
    }

    #[test]
    fn test_module_add_def() -> Result<()> {
        let mut module = IRModule::empty()?;
        let name = GlobalTypeVar::new("my_type", TypeKind::Type, Span::null());
        let type_data = TypeData::new(name.clone(), vec![], vec![], Span::null());
        module.add_def(name.clone(), type_data, true)?;
        let _by_gtv = module.lookup_def(name)?;
        let _by_gv = module.lookup_def_str("my_type")?;
        Ok(())
    }

    #[test]
    fn test_get_global_var() -> Result<()> {
        let mut module = IRModule::empty()?;
        let x = Var::static_tensor("x".into(), vec![1, 1], DataType::float32());
        let params = vec![x.clone()];
        let func = relay::Function::simple(params, x);
        let gv_foo = GlobalVar::new("foo".into(), Span::null());
        let module = module.add(gv_foo.clone(), func)?;
        let gv = module.get_global_var("foo")?;
        assert_eq!(gv_foo, gv);
        Ok(())
    }

    #[test]
    fn test_get_global_vars() -> Result<()> {
        let names = vec!["foo", "bar", "baz"];
        let module = add_dummy_functions(names.clone())?;
        let gvars: Vec<String> = module
            .get_global_vars()?
            .into_iter()
            .map(|gv| gv.name_hint.as_str().unwrap().to_string())
            .collect();

        for name in names {
            assert!(gvars.contains(&name.to_string()));
        }

        Ok(())
    }

    #[test]
    fn test_get_global_type_vars() -> Result<()> {
        let names = vec!["foo", "bar", "baz"];
        let module = add_dummy_types(names.clone())?;
        let gvars: Vec<String> = module
            .get_global_type_vars()?
            .into_iter()
            .map(|gv| gv.name_hint.as_str().unwrap().to_string())
            .collect();

        for name in names {
            assert!(gvars.contains(&name.to_string()));
        }

        Ok(())
    }

    #[test]
    fn test_contains_global_var() -> Result<()> {
        let module = add_dummy_functions(vec!["foo"])?;
        assert!(module.contains_global_var("foo")?);
        Ok(())
    }

    #[test]
    fn test_contains_global_type_var() -> Result<()> {
        let module = add_dummy_types(vec!["foo"])?;
        assert!(module.contains_global_type_var("foo")?);
        Ok(())
    }

    // TODO(@jroesch): not really sure about this API at all.
    // pub fn lookup_tag(&self, tag: i32) -> Result<relay::Constructor> {
    //     module_lookup_tag(self.clone(), tag)
    // }

    #[test]
    fn test_from_expr() -> Result<()> {
        let x = Var::static_tensor("x".into(), vec![1, 1], DataType::float32());
        let params = vec![x.clone()];
        let func = relay::Function::simple(params, x);
        let module = IRModule::from_expr(func.clone())?;
        let main_fn = module.lookup_str("main")?;
        let main_fn = main_fn.downcast::<relay::Function>()?;
        assert_eq!(main_fn, func);
        Ok(())
    }

    #[test]
    fn test_import() -> Result<()> {
        let mut std_path: String = env!("CARGO_MANIFEST_DIR").into();
        std_path += "/../../python/tvm/relay/std/prelude.rly";

        let mut mod1 = IRModule::empty()?;
        mod1.import(std_path.clone())?;
        mod1.lookup_str("map")?;

        // TODO(@jroesch): this requires another patch of mine to enable.

        // if cfg!(feature = "python") {
        //     crate::python::load().unwrap();
        //     let mut mod2 = IRModule::empty()?;
        //     mod2.import_from_std("prelude.rly")?;
        //     mod2.lookup_str("map")?;
        // }

        Ok(())
    }
}

File: python/tvm/ir/module.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""IRModule that holds the functions and type definitions."""

from __future__ import annotations

from typing import Dict, Union

import tvm.ffi
from tvm.runtime import Scriptable
from tvm.runtime.object import Object

from . import _ffi_api
from . import expr as _expr
from .attrs import DictAttrs
from .base import Node


@tvm.ffi.register_object("ir.IRModule")
class IRModule(Node, Scriptable):
    """IRModule that holds functions and type definitions.

    IRModule is the basic unit for all IR transformations across the stack.

    Parameters
    ----------
    functions: Optional[dict].
        Map of global var to BaseFunc
    """

    def __init__(self, functions=None, attrs=None, global_infos=None):
        if functions is None:
            functions = {}
        elif isinstance(functions, dict):
            mapped_funcs = {}
            for k, v in functions.items():
                if isinstance(k, str):
                    k = _expr.GlobalVar(k)
                if not isinstance(k, _expr.GlobalVar):
                    raise TypeError("Expect functions to be Dict[GlobalVar, Function]")
                mapped_funcs[k] = v
            functions = mapped_funcs

        attrs = None if not attrs else attrs
        if attrs is not None:
            attrs = tvm.ir.make_node("ir.DictAttrs", **attrs)
        if global_infos is None:
            global_infos = {}
        self.__init_handle_by_constructor__(
            _ffi_api.IRModule,
            functions,
            attrs,
            global_infos,
        )

    def clone(self) -> "IRModule":
        return _ffi_api.Module_Clone(self)

    def functions_items(self):
        """Get items in self.functions.items() in alphabetical order.

        Returns
        -------
        items: List[Tuple[GlobalVar, Function]]
            The functions items.
        """
        items = list(self.functions.items())
        items.sort(key=lambda item: str(item[0].name_hint))
        return items

    def __setitem__(self, var, val):
        """Add a mapping to the module.

        Parameters
        ---------
        var: GlobalVar
            The global variable.

        val: Union[Function, Type]
            The value.
        """
        return self._add(var, val, True)

    def _add(self, var, val, update=True):
        if isinstance(val, _expr.RelaxExpr):
            if isinstance(var, str):
                if _ffi_api.Module_ContainGlobalVar(self, var):
                    var = _ffi_api.Module_GetGlobalVar(self, var)
                else:
                    var = _expr.GlobalVar(var)
            _ffi_api.Module_Add(self, var, val, update)

    def __getitem__(self, var):
        """Lookup a global definition by name or by variable.

        Parameters
        ----------
        var: Union[String, GlobalVar, GlobalTypeVar]
            The name or global variable.

        Returns
        -------
        val: Union[Function, Type]
            The definition referenced by :code:`var` (either a function or type).
        """
        if isinstance(var, str):
            return _ffi_api.Module_Lookup_str(self, var)
        assert isinstance(var, _expr.GlobalVar)
        return _ffi_api.Module_Lookup(self, var)

    def __delitem__(self, var: Union[str, _expr.GlobalVar]):
        _ffi_api.Module_Remove(self, var)

    def __contains__(self, var: Union[str, _expr.GlobalVar]) -> bool:
        return _ffi_api.Module_Contains(self, var)

    def update(self, other):
        """Insert functions in another Module to current one.

        Parameters
        ----------
        other: IRModule
            The module to merge into the current Module.
        """
        if isinstance(other, dict):
            other = IRModule(other)

        return _ffi_api.Module_Update(self, other)

    def update_func(self, var, func):
        """Update the function corresponding to a global variable in the
        module.

        Parameters
        ----------
        var: GlobalVar
            The global variable.

        func: tvm.ir.BaseFunc
            The function to be inserted.
        """
        return _ffi_api.Module_UpdateFunction(self, var, func)

    def update_global_info(self, name, global_info):
        """Update global info in the module

        Parameters
        ----------
        name: str
            The name for the global info.

        global_info: List[GlobalInfo]
            The global info to be updated.
        """
        return _ffi_api.Module_UpdateGlobalInfo(self, name, global_info)

    def get_global_var(self, name):
        """Get a global variable in the function by name.

        Parameters
        ----------
        name: str
            The name of the global variable.

        Returns
        -------
        global_var: GlobalVar
            The global variable mapped to :code:`name`.

        Raises
        ------
        tvm.error.TVMError if we cannot find corresponding global var.
        """
        return _ffi_api.Module_GetGlobalVar(self, name)

    def get_global_vars(self):
        """Collect all global vars defined in this module.

        Returns
        -------
        global_vars: Array[GlobalVar]
            An array of global vars.
        """
        return _ffi_api.Module_GetGlobalVars(self)

    def replace_global_vars(
        self,
        replacements: Dict[Union[str, _expr.GlobalVar], Union[str, _expr.GlobalVar]],
    ) -> "IRModule":
        """Replace GlobalVar instances within the module

        Replace GlobalVars within the IRModule.  Since the IRModule
        may contain internal references to a GlobalVar, either in TIR
        or in Relax, this method should be used whenever replacing or
        renaming a GlobalVar.

        Parameters
        ----------
        replacements: Dict[Union[str, _expr.GlobalVar], Union[str, _expr.GlobalVar]]

            A dictionary where each key is a GlobalVar to be replaced,
            and the corresponding value is the GlobalVar with which to
            replace it.

        Returns
        -------
        IRModule
            The updated module

        """
        return _ffi_api.Module_ReplaceGlobalVars(self, replacements)

    @staticmethod
    def from_expr(expr, functions=None):
        """Construct a module from a standalone expression.

        Parameters
        ----------
        expr: RelaxExpr
            The starting expression

        global_funcs: Optional[dict]
            Map of global vars to function definitions

        Returns
        -------
        mod: Module
            A module containing the passed definitions,
            where expr is set as the entry point
            (wrapped in a function if necessary)
        """
        funcs = functions if functions is not None else {}
        return _ffi_api.Module_FromExpr(expr, funcs)

    def get_attr(self, attr_key):
        """Get the IRModule attribute.

        Parameters
        ----------
        attr_key : str
            The attribute key.

        Returns
        -------
        attr_value : Any
            Attribute value
        """

        return _ffi_api.Module_GetAttr(self, attr_key)

    def with_attr(self, attr_key, attr_value):
        """Copy the IRModule and add an attribute to it.

        Parameters
        ----------
        attr_key : str
            The attribute key.

        attr_value : Object
            The new attribute value.

        Returns
        -------
        mod : IRModule
            A new copy of the IRModule with the attribute
        """

        return _ffi_api.Module_WithAttr(self, attr_key, attr_value)

    def without_attr(self, attr_key: str) -> "IRModule":
        """Copy the IRModule and remove an attribute key and its associated value.
        Parameters
        ----------
        attr_key : str
            The attribute key.
        Returns
        -------
        mod : IRModule
            A new copy of the IRModule without the attribute
        """

        return _ffi_api.Module_WithoutAttr(self, attr_key)

    def with_attrs(self, attr_map: Union[DictAttrs, Dict[str, Object]]) -> "IRModule":
        """Copy the IRModule and add the given attribute map to it.
        Parameters
        ----------
        attr_map: Union[DictAttrs, Dict[str, Object]]
            The attribute map
        Returns
        -------
        mod : IRModule
            A new copy of the IRModule with the attribute
        """
        if isinstance(attr_map, tvm.ir.DictAttrs):
            attr_map = attr_map._dict()

        return _ffi_api.Module_WithAttrs(self, attr_map)
